authentication/active_directory_authentication_provider.py

```diff
@@ -0,0 +1,53 @@
+#########
+# Copyright (c) 2016 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  * See the License for the specific language governing permissions and
+#  * limitations under the License.
+
+from ldap_authentication_provider import LDAPAuthenticationProvider
+
+DEFAULT_AUTH_ATTR_NAME = 'sAMAccountName'
+
+
+class ActiveDirectoryAuthenticationProvider(LDAPAuthenticationProvider):
+
+    def __init__(self, args, domain_name=None, **kwargs):
+        if 'search_props' in kwargs:
+            if 'user_id_attribute' not in kwargs.get('search_props'):
+                kwargs['search_props']['user_id_attribute'] = \
+                    DEFAULT_AUTH_ATTR_NAME
+        super(ActiveDirectoryAuthenticationProvider, self).__init__(args,
+                                                                    **kwargs)
+        self.domain_name = domain_name
+
+    def authenticate(self, userstore=None):
+        who, password = self.get_creds_from_request()
+
+        if self.is_search_based_authentication():
+            self.authenticate_by_attribute(who, password)
+        else:
+            self.lgr.debug('Attempting to connect using the user id provided: '
+                           '{user_id} and domain name: {domain}'
+                           .format(user_id=who, domain=self.domain_name))
+            self.authenticate_by_user_id(who, password)
+        return who
+
+    # By default, active directory allows for authentication using the UPN
+    # attribute. If search_props not specified, the implementation will assume
+    # that UPN/other default attr is used and append the domain_name to the
+    # user_id in-case domain_name was specified in the configuration.
+    def authenticate_by_user_id(self, who, password):
```

Comment:
we should perform validate_bind outside of this method, since we always have to do it regardless of domain or no domain.
This method should can be removed, and you can just use this instead of calling it:
if self.domain_name:
  who = '{0}@{1}'.format(who, self.domain_name)